framework/validators/EmailValidator.php

```diff
@@ -70,19 +70,34 @@ public function init()
      */
     protected function validateValue($value)
     {
-        // make sure string length is limited to avoid DOS attacks
-        if (!is_string($value) || strlen($value) >= 320) {
+        if (!is_string($value)) {
             $valid = false;
-        } elseif (!preg_match('/^(.*<?)(.*)@(.*?)(>?)$/', $value, $matches)) {
+        } elseif (!preg_match('/^(?P<name>(?:"?([^"]*)"?\s)?)(?:\s+)?(?:(?P<open><?)((?P<local>.+)@(?P<domain>[^>]+))(?P<close>>?))$/i', $value, $matches)) {
             $valid = false;
         } else {
-            $domain = $matches[3];
             if ($this->enableIDN) {
-                $value = $matches[1] . idn_to_ascii($matches[2]) . '@' . idn_to_ascii($domain) . $matches[4];
+                $matches['local'] = idn_to_ascii($matches['local']);
+                $matches['domain'] = idn_to_ascii($matches['domain']);
+                $value = $matches['name'] . $matches['open'] . $matches['local'] . '@' . $matches['domain'] . $matches['close'];
             }
-            $valid = preg_match($this->pattern, $value) || $this->allowName && preg_match($this->fullPattern, $value);
-            if ($valid && $this->checkDNS) {
-                $valid = checkdnsrr($domain, 'MX') || checkdnsrr($domain, 'A');
+
+            if (strlen($matches['local']) > 64 || mb_strlen($matches['name']) > 64) {
```

Comment:
The order of checks here looks a bit weird. Also using default encoding for mb_strlen is not a good idea.